quartz/components/scripts/search.inline.ts

```diff
@@ -64,13 +70,60 @@ function highlight(searchTerm: string, text: string, trim?: boolean) {
       }
       return tok
     })
+    .slice(startIndex, endIndex + 1)
     .join(" ")
 
   return `${startIndex === 0 ? "" : "..."}${slice}${
     endIndex === tokenizedText.length - 1 ? "" : "..."
   }`
 }
 
+function highlightHTML(searchTerm: string, el: HTMLElement) {
+  // try to highlight longest tokens first
+  const p = new DOMParser()
+  const tokenizedTerms = tokenizeTerm(searchTerm)
+  const html = p.parseFromString(el.innerHTML, "text/html")
+
+  const createHighlightSpan = (text: string) => {
+    const span = document.createElement("span")
+    span.className = "highlight"
+    span.textContent = text
+    return span
+  }
+
+  const highlightTextNodes = (node: Node) => {
+    if (node.nodeType === Node.TEXT_NODE) {
+      let nodeText = node.nodeValue || ""
+      tokenizedTerms.forEach((term) => {
+        const escapedTerm = term.replace(/[.*+?^${}()|[\]\\]/g, "\\$&") // Escape special characters
+        const regex = new RegExp(escapedTerm, "gi")
+        if (regex.test(nodeText)) {
+          const matches = nodeText.match(regex)
+          if (matches) {
+            const spanContainer = document.createElement("span")
+            let lastIndex = 0
+            matches.forEach((match) => {
+              const matchIndex = nodeText.indexOf(match, lastIndex)
+              spanContainer.appendChild(
+                document.createTextNode(nodeText.slice(lastIndex, matchIndex)),
+              )
+              spanContainer.appendChild(createHighlightSpan(match))
+              lastIndex = matchIndex + match.length
+            })
+            spanContainer.appendChild(document.createTextNode(nodeText.slice(lastIndex)))
+            node.parentNode?.replaceChild(spanContainer, node)
```

Comment:
does this account for cases where a single text node has multiple matches? (i think so but want to confirm)

Reply:
I believe so.

<img width="640" alt="Screenshot 2024-02-01 at 16 19 28" src="https://github.com/jackyzha0/quartz/assets/29749331/1d7bd1b5-5504-41f3-ba34-5d9895a29bee">